examples/nlp/dialogue_state_tracking/dialogue_state_tracking_sgd.py

```diff
@@ -0,0 +1,456 @@
+            non_cat_slot_status_mask=non_cat_slot_status_mask,
+            logit_noncat_slot_start=logit_noncat_slot_start,
+            logit_noncat_slot_end=logit_noncat_slot_end,
+            noncategorical_slot_value_start=data.noncategorical_slot_value_start,
+            noncategorical_slot_value_end=data.noncategorical_slot_value_end,
+        )
+        tensors = [loss]
+    else:
+        tensors = [
+            data.example_id_num,
+            data.service_id,
+            data.is_real_example,
+            data.start_char_idx,
+            data.end_char_idx,
+            logit_intent_status,
+            logit_req_slot_status,
+            logit_cat_slot_status,
+            logit_cat_slot_value,
+            logit_noncat_slot_status,
+            logit_noncat_slot_start,
+            logit_noncat_slot_end,
+            data.intent_status_labels,
+            data.requested_slot_status,
+            data.categorical_slot_status,
+            data.num_categorical_slots,
+            data.categorical_slot_values,
+            data.noncategorical_slot_status,
+            data.num_noncategorical_slots,
+        ]
+
+    steps_per_epoch = math.ceil(len(datalayer) / (args.train_batch_size * args.num_gpus))
+    return steps_per_epoch, tensors
+
+
+steps_per_epoch, train_tensors = create_pipeline()
+logging.info(f'Steps per epoch: {steps_per_epoch}')
+_, eval_tensors = create_pipeline(dataset_split=args.eval_dataset)
+
+# Create trainer and execute training action
+train_callback = nemo.core.SimpleLossLoggerCallback(
+    tensors=train_tensors,
+    print_func=lambda x: logging.info("Loss: {:.8f}".format(x[0].item())),
+    get_tb_values=lambda x: [["loss", x[0]]],
+    tb_writer=nf.tb_writer,
+    step_freq=args.loss_log_freq if args.loss_log_freq > 0 else steps_per_epoch,
+)
+
+# we'll write predictions to file in DSTC8 format during evaluation callback
+input_json_files = [
+    os.path.join(args.data_dir, args.eval_dataset, 'dialogues_{:03d}.json'.format(fid))
+    for fid in data_processor.FILE_RANGES[args.task_name][args.eval_dataset]
+]
+schema_json_file = os.path.join(args.data_dir, args.eval_dataset, 'schema.json')
+
+# Write predictions to file in DSTC8 format.
+prediction_dir = os.path.join(nf.work_dir, 'predictions', 'pred_res_{}_{}'.format(args.eval_dataset, args.task_name))
+output_metric_file = os.path.join(nf.work_dir, 'metrics.txt')
+os.makedirs(prediction_dir, exist_ok=True)
+
+eval_callback = nemo.core.EvaluatorCallback(
+    eval_tensors=eval_tensors,
+    user_iter_callback=lambda x, y: eval_iter_callback(x, y, schema_preprocessor, args.eval_dataset),
+    user_epochs_done_callback=lambda x: eval_epochs_done_callback(
+        x,
+        input_json_files,
+        args.eval_dataset,
+        args.data_dir,
+        prediction_dir,
+        output_metric_file,
+        args.state_tracker,
+        args.debug_mode,
+        schema_preprocessor,
+        args.joint_acc_across_turn,
+        args.no_fuzzy_match,
+    ),
```

Comment:
number of things that you need to pass to that function surprises me every time :]